### PR TITLE
feat(cpu-utils): introduced new abstraction for pinning and topology fetching

### PR DESCRIPTION
**Pull request by gurinderu**

### CODE REVIEW

File: crates/cpu-utils/src/pinning.rs [hunk 1]
```diff
@@ -18,23 +18,49 @@ use ccp_core_affinity::CoreId;
...
+pub struct DefaultThreadPinner;
+
+impl ThreadPinner for DefaultThreadPinner {
+    #[inline]
+    fn pin_current_thread_to(&self, core_id: LogicalCoreId) -> bool {
+        let core_id = CoreId { id: core_id.into() };
+        ccp_core_affinity::set_for_current(core_id)
+    }
+
+    #[inline]
```
Comment (monoid) [inline crates/cpu-utils/src/pinning.rs:40]: I don't think inline will be useful here.

File: default.nix
Comment (monoid) [inline default.nix:1]: It is unexpected.  It may be useful for your convenience, but who will maintain it?

File: ccp/src/cu/cu_prover.rs [hunk 2]
```diff
@@ -23,7 +23,7 @@ use ccp_randomx::Dataset;
 use ccp_randomx::RandomXFlags;
 use ccp_shared::types::*;
 use ccp_utils::run_utils::run_unordered;
-use cpu_utils::CPUTopology;
+use cpu_utils::{CPUTopology, HwlocCPUTopology};
```
Comment (mikevoronov) [inline ccp/src/cu/cu_prover.rs:26]: Let's follow our coding standard and flatten them.

File: default.nix
Comment (gurinderu) [reply default.nix:1]: Done

File: ccp/src/cu/proving_thread_utils/allocator.rs [hunk 3]
```diff
@@ -20,7 +20,7 @@ use ccp_config::ThreadsPerCoreAllocationPolicy;
 use ccp_msr::MSRModeEnforcer;
 use ccp_shared::types::LogicalCoreId;
 use ccp_shared::types::PhysicalCoreId;
-use cpu_utils::CPUTopology;
+use cpu_utils::{CPUTopology, HwlocCPUTopology};
```
[the remark this answers is not in the document]
Comment (gurinderu) [reply ccp/src/cu/proving_thread_utils/allocator.rs]: Fixed

File: crates/cpu-utils/src/pinning.rs [hunk 4]
```diff
@@ -18,23 +18,49 @@ use ccp_core_affinity::CoreId;
 
 use crate::LogicalCoreId;
 
-/// Lightweight function which doesn't require topology to pin current thread to the specified core.
-/// Returns true, if pinning was successful.
-pub fn pin_current_thread_to(core_id: LogicalCoreId) -> bool {
-    let core_id = CoreId { id: core_id.into() };
-    ccp_core_affinity::set_for_current(core_id)
+pub const DUMMY: DummyThreadPinner = DummyThreadPinner {};
```
[the remark this answers is not in the document]
Comment (gurinderu) [reply crates/cpu-utils/src/pinning.rs]: Done

File: crates/cpu-utils/Cargo.toml [hunk 5]
```diff
@@ -12,11 +12,14 @@ publish = true
...
+
 [dependencies]
 ccp-shared.workspace = true
-
 ccp_core_affinity.workspace = true
 hwlocality.workspace = true
 nonempty.workspace = true
 serde.workspace = true
 thiserror.workspace = true
+mockall = { version = "0.12.1", optional = true }
```
[the remark this answers is not in the document]
Comment (gurinderu) [reply crates/cpu-utils/Cargo.toml]: Done

File: ccp/src/cu/cu_prover.rs (hunk 2 above)
Comment (gurinderu) [reply ccp/src/cu/cu_prover.rs]: Fixed

File: crates/cpu-utils/src/pinning.rs (hunk 1 above)
Comment (gurinderu) [reply crates/cpu-utils/src/pinning.rs:38]: it is not a big deal)

File: crates/cpu-utils/src/pinning.rs [hunk 6]
```diff
@@ -17,24 +17,34 @@
 use ccp_core_affinity::CoreId;
 
 use crate::LogicalCoreId;
```
Comment (mikevoronov) [inline crates/cpu-utils/src/pinning.rs:19]: add newline please

**Review by mikevoronov — APPROVED** [review]
utACK

**Pull request merged**